projectriff/node-function-invoker
# Execute hooks once per process, not invocation

Fixes #146.

## How to validate

1. create any of the variants of the function, e.g. `/tmp/captain_hook.js` (see missing code for each variant below):
```js
const getRandomInt = (max) => {
        return Math.floor(Math.random() * Math.floor(max));
};
let counter = 0;
module.exports = () => {
        return counter++;
};
// see hook variants below
```
2. point the node invoker to the PR branch
3. build (`make`) and run the [adapter](https://github.com/projectriff/streaming-http-adapter)
```shell
FUNCTION_URI=/tmp/captain_hook.js NODE_DEBUG='riff' ./streaming-http-adapter node ~/workspace/node-function-invoker/server.js
```
The server should shut down and you should observe the corresponding function logs.

## `captain_hook.js` variants

Note that I did not mention the explicit use of `Promise` in the below `async` sections, only keep things brief. You can also try that.

### `$init` fails

When the `$init` hook fails, the server should shut down (forceful shutdown in that case) and the `$destroy` hook should be executed.
The process status should be 1.

#### async
```js
module.exports.$init = async () => {
        throw new Error('init KO');
};
module.exports.$destroy = () => { // can be async or not, does not matter here
        console.log(`Destroying, counter was ${counter}`);
        counter = -1;
};
```

#### sync
```js
module.exports.$init = () => {
        throw new Error('init KO');
};
module.exports.$destroy = () => { // can be async or not, does not matter here
        console.log(`Destroying, counter was ${counter}`);
        counter = -1;
};
```

### `$destroy` fails

Run as many `curl http://localhost:8080 -XPOST -H'Accept:text/plain' -v` as you want.
Then, `Ctrl-c` the active server process to trigger the graceful shutdown and the `$destroy` hook execution.
The process status should be 1.

#### async
```js
module.exports.$init = () => { // can be async or not, does not matter here
        counter = getRandomInt(42);
        console.log(`Starting with ${counter}`);
};
module.exports.$destroy = async () => {
        throw new Error('destroy KO');
};
```

#### sync
```js
module.exports.$init = async () => { // can be async or not, does not matter here
        counter = getRandomInt(42);
        console.log(`Starting with ${counter}`);
};
module.exports.$destroy = () => {
        throw new Error('destroy KO');
};
```

### `$init` and `$destroy` fails

You should see logs of the failure of both hooks.
You can also try other combinations (async init/sync destroy and vice-versa).
The process status should be 1.

#### async
```js
module.exports.$init = async () => {
        throw new Error('init KO');
};
module.exports.$destroy = async () => {
        throw new Error('destroy KO');
};
```

#### sync
```js
module.exports.$init = () => {
        throw new Error('init KO');
};
module.exports.$destroy = () => {
        throw new Error('destroy KO');
};
```
